require('./test');

var assert = require("assert");
require("./assert-helper");

var path = require('path');
var wait = require('wait.for');

var FileModuleStore = require("../lib/file-module-store");

var logger = require('log4js').getLogger("file-module-store-test");

describe("File module store", function()
{
    it("should list containers", function(done) 
    {
    	wait.launchFiber(function()
    	{
	    	var moduleStore = new FileModuleStore({ directory: __dirname });

	    	var containers = moduleStore.getAppContainers();
	    	var expectedContainers = ["testapp"];

	    	assert.deepEqual(containers, expectedContainers);
	    	done();
	    });
    });

    it("should create container", function(done) 
    {
    	wait.launchFiber(function()
    	{
	    	var moduleStore = new FileModuleStore({ directory: __dirname });

	    	moduleStore.createAppContainer("foo");

	    	var containers = moduleStore.getAppContainers();
	    	var expectedContainers = ["testapp", "foo"];

	    	assert.deepEqual(containers.sort(), expectedContainers.sort());
	    	done();
	    });
    });

    it("should delete container", function(done) 
    {
        var moduleStore = new FileModuleStore({ directory: __dirname });
        
        wait.launchFiber(function ()
        {
            moduleStore.deleteAppContainer("foo");
        });
         
        // I get the following error fairly frequently when I run this test...
        //
        // Uncaught Error: EPERM: operation not permitted, stat 'C:\Dev\SynchroServer\SynchroServer\node_modules\synchro-api\test\foo'
        //   at Error(native)            
        //
        // What is happening is that in getAppContainers() it does an fs.readdir() which sometimes returns the very recently 
        // deleted "foo" directory, then when it iterates the found directories and does an isDirectory() on "foo", you get the
        // error (because foo actually isn't there).
        //
        // All of the file operations (the delete, the readdir, and the isDirectory/stat) are all async and awaited properly, so
        // I don't think this is our fault. 
        //
        // I added the wait below.  Even with values of up to 500ms, I still saw relatively frequent failures.  I had to crank it all
        // to the way to 1000ms to get 10 in a row without a failure. 
        //            
        setTimeout(function ()
        {
            wait.launchFiber(function ()
            {
                var containers = moduleStore.getAppContainers();
                var expectedContainers = ["testapp"];
                
                assert.deepEqual(containers, expectedContainers);
                done();
            });
	    }, 1000);
    });

    it("should fail to create container that already exists", function(done) 
    {
    	wait.launchFiber(function()
    	{
	    	var moduleStore = new FileModuleStore({ directory: __dirname });

	    	try
	    	{
		    	moduleStore.createAppContainer("testapp");
	    	}
	    	catch (err)
	    	{
		    	done();
	    	}
	    });
    });

    it("should fail to delete container that doesn't exist", function(done) 
    {
    	wait.launchFiber(function()
    	{
	    	var moduleStore = new FileModuleStore({ directory: __dirname });

	    	try
	    	{
		    	moduleStore.deleteAppContainer("foo");
	    	}
	    	catch (err)
	    	{
		    	done();
	    	}
	    });
    });

    it("should fail to delete container that doesn't contain app", function(done) 
    {
    	wait.launchFiber(function()
    	{
	    	var moduleStore = new FileModuleStore({ directory: __dirname });

	    	try
	    	{
		    	moduleStore.deleteAppContainer("testconfig");
	    	}
	    	catch (err)
	    	{
		    	done();
	    	}
	    });
    });

    it("should get app module store from valid container", function(done) 
    {
    	wait.launchFiber(function()
    	{
	    	var moduleStore = new FileModuleStore({ directory: __dirname });

	    	var appModuleStore = moduleStore.getAppModuleStore("testapp");

	    	var app = appModuleStore.getAppDefinition();

	    	assert.equal(app.name, "synchro-test");
	    	assert.equal(app.version, "0.0.0");
	    	done();
	    });
    });

    it("should fail to get app module store from invalid valid container", function(done) 
    {
    	wait.launchFiber(function()
    	{
	    	var moduleStore = new FileModuleStore({ directory: __dirname });

	    	try
	    	{
		    	moduleStore.getAppModuleStore("foo");
	    	}
	    	catch (err)
	    	{
		    	done();	    		
	    	}
	    });
    });
});
